app/domain/authentication/authn_gcp/extract_resource_restrictions.rb

```diff
@@ -0,0 +1,88 @@
+require 'command_class'
+
+module Authentication
+  module AuthnGCP
+
+    # This class is responsible of restrictions extraction that are set on a Conjur host or user as annotations.
+    ExtractResourceRestrictions = CommandClass.new(
+      dependencies: {
+        role_class:     ::Role,
+        resource_class: ::Resource,
+        logger:         Rails.logger
+      },
+      inputs:       %i(account username extraction_prefix)
+    ) do
+
+      def call
+        extract_resource_restrictions
+        restrictions_list
+      end
+
+      private
+
+      def extract_resource_restrictions
+        @logger.debug(LogMessages::Authentication::AuthnGCP::ExtractingRestrictionsFromResource.new(@username, @extraction_prefix))
+        prefixed_resource_annotations
+        init_restrictions_list
+        @logger.debug(LogMessages::Authentication::AuthnGCP::ExtractedResourceRestrictions.new(restrictions_list.length()))
+      end
+
+      def restrictions_list
+        return @restrictions_list if @restrictions_list
+
+        @restrictions_list = Array.new
+      end
+
+      def prefixed_resource_annotations
+        @prefixed_resource_annotations ||= resource_annotations.select do |a|
+          annotation_name = a.values[:name]
+          annotation_name.start_with?(@extraction_prefix)
+        end
+      end
+
+      def init_restrictions_list
+        prefixed_resource_annotations.select do |a|
+          annotation_name = a.values[:name]
+          resource_value = annotation_value(annotation_name)
+          next unless resource_value
+          restrictions_list.push(
+            ResourceRestriction.new(
+              type: annotation_name,
```

Comment:
`annotation_name` will hold also the prefix. do you want it to be that way?

Reply:
Yes, for now I don’t see a good reason to delete the prefix

Reply:
i think you will need to once you validate them but i guess it's ok. it's quite odd to have a `ResourceRestriction`:
```
type: authn-gcp/project_id
value: some_project_id`
```
but i guess we can live with it.

by the way, in this implementation we may have a `ResourceRestriction`:
```
type: authn-gcp/blah
value: some_blah`
```

this will be removed in the `ValidateResourceRestrictions` stage? we first populate all the values and only then validate them? another approach can be to populate a value only if it's validated

Reply:
Yes my idea was first to populate all the values without any validation, so this class will only one responsibility and may be useful later in other places.
And in the `ValidateResourceRestrictions` stage we will validate the values.

Reply:
i'm ok with that approach. we can challenge it once we implement `ValidateResourceRestrictions`